lib/opam_solve.ml

```diff
@@ -201,10 +221,15 @@ module Opam_monorepo_context (Base_context : BASE_CONTEXT) :
           require_cross_compile
           && List.exists ~f:candidate_cross_compile candidates
         in
+        let preferred_version =
+          let open Option.O in
+          let* preferred_versions = preferred_versions in
+          OpamPackage.Name.Map.find_opt name preferred_versions
+        in
         filter_candidates ~allow_jbuilder ~must_cross_compile ~require_dune
           ~name candidates
         |> remove_opam_provided ~opam_provided
-        |> demote_candidates_to_avoid
+        |> sort_candidates_by_preference ~preferred_version
```

Comment:
Is there a way to keep this as two steps, somehow? I feel like the double processing overhead is outweighted by the readability of having functions do one thing. Then `sort_candidates_by_preference` becomes basically just a quick pattern match.